pkg/gardenlet/controller/managedseed/managedseed_actuator.go

```diff
@@ -0,0 +1,636 @@
+	gardenNamespace := &corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: v1beta1constants.GardenNamespace}}
+	if err := shootClient.Client().Create(ctx, gardenNamespace); err != nil && !apierrors.IsAlreadyExists(err) {
+		return fmt.Errorf("could not create garden namespace in shoot %q: %+v", kutil.ObjectName(shoot), err)
+	}
+
+	switch {
+	case managedSeed.Spec.SeedTemplate != nil:
+		// Register the shoot as seed
+		managedSeedLogger.Infof("Registering shoot %s as seed", kutil.ObjectName(shoot))
+		if err := a.registerAsSeed(ctx, managedSeed, shoot); err != nil {
+			return fmt.Errorf("could not register shoot %q as seed: %+v", kutil.ObjectName(shoot), err)
+		}
+
+	case managedSeed.Spec.Gardenlet != nil:
+		// Deploy gardenlet into the shoot, it will register the seed automatically
+		managedSeedLogger.Infof("Deploying gardenlet into shoot %s", kutil.ObjectName(shoot))
+		if err := a.deployGardenlet(ctx, shootClient, managedSeed, shoot); err != nil {
+			return fmt.Errorf("could not deploy gardenlet into shoot %q: %+v", kutil.ObjectName(shoot), err)
+		}
+	}
+
+	return nil
+}
+
+// Delete reconciles ManagedSeed deletion.
+func (a *actuator) Delete(ctx context.Context, managedSeed *seedmanagementv1alpha1.ManagedSeed, shoot *gardencorev1beta1.Shoot) error {
+	managedSeedLogger := logger.NewFieldLogger(a.logger, "managedSeed", kutil.ObjectName(managedSeed))
+
+	// Get shoot client
+	shootClient, err := a.clientMap.GetClient(ctx, keys.ForShoot(shoot))
+	if err != nil {
+		return fmt.Errorf("could not get shoot client for shoot %q: %+v", kutil.ObjectName(shoot), err)
+	}
+
+	switch {
+	case managedSeed.Spec.SeedTemplate != nil:
+		// Unregister the shoot as seed
+		managedSeedLogger.Infof("Unregistering shoot %s as seed", kutil.ObjectName(shoot))
+		if err := a.unregisterAsSeed(ctx, managedSeed); err != nil {
+			return fmt.Errorf("could not unregister shoot %q as seed: %+v", kutil.ObjectName(shoot), err)
+		}
+
+	case managedSeed.Spec.Gardenlet != nil:
+		// Ensure the seed is deleted
+		managedSeedLogger.Infof("Ensuring seed %s is deleted", managedSeed.Name)
+		if err := a.ensureSeedDeleted(ctx, managedSeed); err != nil {
+			return fmt.Errorf("could not ensure seed %q is deleted: %+v", managedSeed.Name, err)
+		}
+
+		// Delete gardenlet from the shoot
+		managedSeedLogger.Infof("Deleting gardenlet from shoot %s", kutil.ObjectName(shoot))
+		if err := a.deleteGardenlet(ctx, shootClient, managedSeed, shoot); err != nil {
+			return fmt.Errorf("could not delete gardenlet from shoot %q: %+v", kutil.ObjectName(shoot), err)
+		}
+	}
+
+	// Delete garden namespace from the shoot
+	managedSeedLogger.Infof("Deleting garden namespace from shoot %s", kutil.ObjectName(shoot))
+	gardenNamespace := &corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: v1beta1constants.GardenNamespace}}
+	if err := shootClient.Client().Delete(ctx, gardenNamespace); client.IgnoreNotFound(err) != nil {
```

Comment:
The old implementation didn't delete the `garden` namespace at all, so breaking behavioral compatibility is not an issue - it's already heavily broken, but hopefully for the better.

My thinking here is similar to https://github.com/gardener/gardener/pull/3418#discussion_r573522642 - we could improve the UX by waiting for the garden namespace to be deleted, but the impact is quite low and therefore I wouldn't make things more complex just because of this. Or do you see perhaps a bigger impact in this case?

Reply:
We agreed it's rather important to ensure that the namespace has been properly deleted before removing the finalizer from the managed seed resource (and allowing it to be deleted). The impact is indeed bigger in this case since if the `garden` namespace is stuck in deletion, the shoot can't be registered as seed again. Will do as suggested.

Reply:
Done. I also decided to implement a similar logic for the seed secrets - ensure that they are actually deleted rather than just triggering the deletion. So the deletion flow behaves consistently for all owned objects:

* Deletes the seed and ensures that it's actually deleted
* Deletes the seed secrets (if they have been previously created) and ensures they are actually deleted
* Deletes the garden namespace and ensures it's actually deleted